azurelinuxagent/ga/extension_telemetry.py

```diff
@@ -0,0 +1,402 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import datetime
+import json
+import os
+import re
+import threading
+from collections import defaultdict
+
+import azurelinuxagent.common.logger as logger
+from azurelinuxagent.common import conf
+from azurelinuxagent.common.event import EVENTS_DIRECTORY, TELEMETRY_LOG_EVENT_ID, \
+    TELEMETRY_LOG_PROVIDER_ID, add_event, WALAEventOperation, add_log_event, get_event_logger
+from azurelinuxagent.common.exception import InvalidExtensionEventError
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.telemetryevent import TelemetryEventList, TelemetryEvent, TelemetryEventParam, \
+    GuestAgentGenericLogsSchema
+from azurelinuxagent.ga.exthandlers import HANDLER_NAME_PATTERN
+from azurelinuxagent.ga.periodic_operation import PeriodicOperation
+
+
+def get_extension_telemetry_handler(protocol_util):
+    return ExtensionTelemetryHandler(protocol_util)
+
+class ExtensionEventSchema(object):
+    """
+    Class for defining the schema for Extension Events.
+    """
+    Version = "Version"
+    Timestamp = "Timestamp"
+    TaskName = "TaskName"
+    EventLevel = "EventLevel"
+    Message = "Message"
+    EventPid = "EventPid"
+    EventTid = "EventTid"
+    OperationId = "OperationId"
+
+class ProcessExtensionTelemetry(PeriodicOperation):
+    """
+    Periodic operation for collecting and sending extension telemetry events to Wireserver.
+    """
+
+    _EXTENSION_EVENT_COLLECTION_PERIOD = datetime.timedelta(minutes=5)
+    _EXTENSION_EVENT_FILE_NAME_REGEX = re.compile(r"^(\d+)\.json$", re.IGNORECASE)
+
+    # Limits
+    _MAX_NUMBER_OF_EVENTS_PER_EXTENSION_PER_PERIOD = 300
+    _EXTENSION_EVENT_FILE_MAX_SIZE = 4 * 1024 * 1024  # 4 MB = 4 * 1,048,576 Bytes
+    _EXTENSION_EVENT_MAX_SIZE = 1024 * 6  # 6Kb or 6144 characters. Limit for the whole event. Prevent oversized events.
+    _EXTENSION_EVENT_MAX_MSG_LEN = 1024 * 3  # 3Kb or 3072 chars.
+
+    _EXTENSION_EVENT_REQUIRED_FIELDS = [attr.lower() for attr in dir(ExtensionEventSchema) if
+                                        not callable(getattr(ExtensionEventSchema, attr)) and not attr.startswith("__")]
```

Comment:
This seems a bit hacky; I know `enum.Enum` is backported past 2.6, would that be a good fit here?
((https://www.geeksforgeeks.org/enum-in-python/))

Reply:
Totally agree on the hackiness but couldn't figure out a better way to do it unfortunately :/
The backport [enum34](https://pypi.org/project/enum34/) is not a standard package and needs to be installed separately from pip. Since this is running on a customer's VM, we cant always rely on external packages to be there. 

That being said, I am experimenting with changing our setup.py to read from requirements.txt (as it should already have been) and seeing it that bundles the requirements as part of the .egg itself. If that works, I'll replace the hack with an Enum as you suggested.